pkg/controller/seed-controller-manager/kubernetes/resources.go

```diff
@@ -1053,3 +1059,36 @@ func hostnameToIPList(ctx context.Context, hostname string) ([]net.IP, error) {
 
 	return ipList, nil
 }
+
+func (r *Reconciler) ensureAuditWebhook(ctx context.Context, c *kubermaticv1.Cluster, data *resources.TemplateData) error {
+	webhookBackendSecret := &corev1.Secret{}
+	err := r.Get(ctx, types.NamespacedName{Name: resources.AuditWebhookSecretName, Namespace: c.Status.NamespaceName}, webhookBackendSecret)
+	// create the secret with a sample audit webhook backend config, which KKP admin can later update if not enforced in the datacenter.
+	if apierrors.IsNotFound(err) {
+		creators := []reconciling.NamedSecretReconcilerFactory{apiserver.AuditWebhookSecretReconciler(data)}
+		if err := reconciling.ReconcileSecrets(ctx, creators, c.Status.NamespaceName, r.Client); err != nil {
+			return nil
+		}
+	}
+	// Reconcile audit webhook secret from dc if audit webhook backend is enforced and secret is specified.
+	if data.DC().Spec.EnforceAuditWebhook.Enabled {
+		secret := &corev1.Secret{}
+		err := r.Get(ctx, types.NamespacedName{Name: data.DC().Spec.EnforceAuditWebhook.AuditWebhook.Name, Namespace: data.DC().Spec.EnforceAuditWebhook.AuditWebhook.Namespace}, secret)
+		if err != nil {
+			return err
+		} else {
+			tempSecret := secret.DeepCopy()
```

Comment:
What's this DeepCopy good for?

In general I think this function can be improved. The current workflow is

1. reconcile a Secret
2. read some data
3. update the Secret

When reconciling (step 1) should already update the Secret. Doing _both_ reconciling _and_ updating of the same Secret feels like misusing the reconciling framework. If you're reconciling, you should not concern yourself with Create/Update operations, that's the whole idea :grin: 

Maybe a better approach would be to do

```go
func ensureAuditWebhook() {
   data = defaultData

  if dc.enforced {
     clusterwideSecret = getClusterwideSecret
     data = clusterwideSecret.Data
  }

  reconciling.ReconcileSecrets(clusterSecret)
}
```

Reply:
Updated, thanks.